src/stim/util_top/reference_sample_tree.inl

```diff
@@ -0,0 +1,161 @@
+}
+
+static uint64_t max_feedback_lookback_in_loop(const Circuit &loop) {
+    uint64_t furthest_lookback = 0;
+    for (const auto &inst : loop.operations) {
+        if (inst.gate_type == GateType::REPEAT) {
+            furthest_lookback = std::max(furthest_lookback, max_feedback_lookback_in_loop(inst.repeat_block_body(loop)));
+        } else {
+            auto f = GATE_DATA[inst.gate_type].flags;
+            if ((f & GateFlags::GATE_CAN_TARGET_BITS) && (f & GateFlags::GATE_TARGETS_PAIRS)) {
+                // Feedback-capable operation. Check for any measurement record targets.
+                for (auto t : inst.targets) {
+                    if (t.is_measurement_record_target()) {
+                        furthest_lookback = std::max(furthest_lookback, (uint64_t)-t.rec_offset());
+                    }
+                }
+            }
+        }
+    }
+    return furthest_lookback;
+}
+
+template <size_t W>
+ReferenceSampleTree CompressedReferenceSampleHelper<W>::do_loop_with_tortoise_hare_folding(const Circuit &loop, uint64_t reps) {
+    if (reps < 10) {
+        return do_loop_with_no_folding(loop, reps);
+    }
+
+    ReferenceSampleTree result;
+    result.repetitions = 1;
+
+    CompressedReferenceSampleHelper<W> tortoise(sim);
+    CompressedReferenceSampleHelper<W> hare(std::move(sim));
+    uint64_t max_feedback_lookback = max_feedback_lookback_in_loop(loop);
+    uint64_t tortoise_steps = 0;
+    uint64_t hare_steps = 0;
+    while (hare_steps < reps) {
+        hare_steps++;
+        result.suffix_children.push_back(hare.do_loop_with_no_folding(loop, 1));
+        assert(result.suffix_children.size() == hare_steps);
+
+        if (tortoise.in_same_recent_state_as(hare, max_feedback_lookback, hare_steps < 10)) {
+            break;
+        }
+
+        // Tortoise advances half as quickly.
+        if (hare_steps & 1) {
+            tortoise_steps++;
+            tortoise.do_loop_with_no_folding(loop, 1);
+        }
+    }
+
+    if (hare_steps == reps) {
+        // No loop found.
+        sim = std::move(hare.sim);
+        return result;
+    }
+
+    // Advance until the remaining iterations are a multiple of the period.
+    assert(result.suffix_children.size() == hare_steps);
+    uint64_t period = hare_steps - tortoise_steps;
+    size_t period_steps_left = (reps - hare_steps) / period;
+    while ((reps - hare_steps) % period) {
+        result.suffix_children.push_back(hare.do_loop_with_no_folding(loop, 1));
+        hare_steps += 1;
+    }
+    assert(hare_steps + period_steps_left * period == reps);
+    assert(hare_steps >= period);
+    sim = std::move(hare.sim);
+
+    // Move the periodic measurements out of the hare's tail, into a loop node.
+    ReferenceSampleTree loop_contents;
+    for (size_t k = hare_steps - period; k < hare_steps; k++) {
+        loop_contents.suffix_children.push_back(std::move(result.suffix_children[k]));
+    }
+    result.suffix_children.resize(hare_steps - period);
```

Comment:
If this `resize()` was before the loop, there'd be no extra alloc and copy.

Reply:
Wouldn't it also lose the contents that are being moved to the other vector before moving them?

Reply:
Ah, I misread which vector was resized. Nevermind.